File: Sliding window/LET_no1074_Number of Submatrices That Sum to Target.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size(), res = 0;
        for (auto &row : matrix)
            for (int j = 0; j < n - 1; j++)
                row[j + 1] += row[j];
        
        unordered_map<int, int> counter;
        for (int j1 = 0; j1 < n; j1++) {
            for (int j2 = j1; j2 < n; j2++) {
                counter = {{0, 1}};
                int cur = 0;
                for (int i = 0; i < m; i++) {
                    cur += matrix[i][j2] - (j1 > 0 ? matrix[i][j1 - 1] : 0);
                    res += counter.find(cur - target) != counter.end() ? counter[cur - target] : 0;
                    counter[cur]++;
                }
            }
        }
        
        return res;
    }
};

File: Sliding window/LET_no1074_Number of Submatrices That Sum to Target.py
class Solution:
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        m, n = len(matrix), len(matrix[0])
        for row in matrix:
            for j in range(n - 1):
                row[j + 1] += row[j]
        
        res = 0
        for j1 in range(n):
            for j2 in range(j1, n):
                pre_sum = collections.defaultdict(int)
                pre_sum[0] = 1
                cur = 0
                for i in range(m):
                    cur += matrix[i][j2] - (matrix[i][j1 - 1] if j1 > 0 else 0)
                    res += pre_sum[cur - target]
                    pre_sum[cur] += 1
        
        return res